mozilla-conduit/lando-api
# Add API endpoint to get a Phabricator Revision from Lando

(Note: this is a larger PR and it may help to view the changes for each commit separately)

### Refactor tests to reduce duplication

The stubs to create the app and versionfile in tests/test_app.py would also be needed in every other test file that wanted to test the Flask app. This commit extracts those utility functions into their own file which can be imported into any test that needs them. This commit also renames 'test_app.py' to 'tests_dockerflow.py' to better represent what that file is testing.
This is based on what David Walsh did in mozilla-conduit/lando-ui, thanks!

### Add API enpoint to query a revision using the new Phabricator Client.

This commit adds a utility class found in phabricator_client.py that abstracts the interface to Phabricator's Conduit API. Currently it supports querying revisions, users, and repos. As Phabricator's API changes, we can take care of those changes in this class without having to rewrite all of our other methods and api endpoints.

Using the newly created class, this commit also adds a new api endpoint `/revisions/{id}` which returns details about a revision given an id. See the updated spec/swagger.yml for more details about this endpoint.

### Add tests for the PhabricatorClient and /revisions/{id} endpoint

This commit adds unit tests for the PhabricatorClient and /revisions/{id} endpoint using the requests_mock library to stub out Phabricator's API responses.

A new folder, `canned_responses` is added to contain example responses returned by the different services we depend on in our unit tests.

## How to test this PR

- First run `docker-compose up --build` to refresh your images.
- You can run `invoke test` to test the updated tests.
- Then run `docker-compose up` again to start the API server. Once running, visit `your-docker-ip:8888` which will take you to the OpenAPI index page which allows you to test out the api directly. Currently, the project is setup to query the Phabricator instance at https://mozphab.dev.mozaws.net/. You can test querying https://mozphab.dev.mozaws.net/D1 by inputting `1` for the id field and your phabricator API key (on the mozphab.dev instance). Inspect the result to confirm it is all working properly.

![lando_rev_api](https://cloud.githubusercontent.com/assets/7828780/26635543/13b47434-45e8-11e7-905b-f7415b953b77.PNG)

(This PR is now complete)
